src/commands/kick.cpp

```diff
@@ -6,15 +6,138 @@
 /*   By: faboussa <faboussa@student.42lyon.fr>      +#+  +:+       +#+        */
 /*                                                +#+#+#+#+#+   +#+           */
 /*   Created: 2024/11/04 10:20:03 by yusengok          #+#    #+#             */
-/*   Updated: 2024/11/12 11:53:47 by faboussa         ###   ########.fr       */
+/*   Updated: 2024/11/13 19:16:56 by faboussa         ###   ########.fr       */
 /*                                                                            */
 /* ************************************************************************** */
 
+#include <iostream>
+#include <sstream>
+#include <string>
+
 #include "../../includes/Server.hpp"
 
-void Server::kick(int fd, const std::string &arg) {
-/*--- To remove ---*/
-  (void)fd;
-  (void)arg;
-/*-----------------*/
+// kick #channelName #targetNick #reason (reason is optional)
+
+void Server::kick(int fd, const std::string &param) {
+  Client &client = _clients.at(fd);
+  if (param.empty()) {
+    send461NeedMoreParams(client, "KICK");
+    return;
+  }
+  std::istringstream iss(param);
+  std::string channelName, targetNick, reason;
+  std::getline(iss, channelName, ' ');
+  std::getline(iss, targetNick, ' ');
+  if (targetNick.empty()) {
+    send461NeedMoreParams(client, "KICK");
+    return;
+  }
+  // Lire la raison si elle existe
+  if (iss.rdbuf()->in_avail() > 0) {
+    std::getline(iss >> std::ws, reason);
+  }
+  if (reason.length() >= gConfig->getLimit(KICKLEN)) {
+    reason = reason.substr(0, gConfig->getLimit(KICKLEN));
+  }
+  if (reason[0] == ':') {
+    reason = reason.substr(1);
+  }
+#ifdef DEBUG
+  std::cout << "[KICK] Channel: " << channelName << " / Target: " << targetNick
+            << " / Reason: " << reason << std::endl;
+#endif
+  if (!channelExists(channelName)) {
+    send403NoSuchChannel(client, channelName);
+    return;
+  }
+  Channel &channel = _channels.at(channelName.substr(1));
+  if (!channel.isOperator(client.getFd())) {
+    send482ChanOPrivsNeeded(client, channel);
+    return;
+  }
+  Client *targetClient = findClientByNickname(targetNick);
+  if (targetClient == NULL) {
+    send401NoSuchNick(client, targetNick);
+    return;
+  }
+  int fdTarget = targetClient->getFd();
+  if (!channel.isClientInChannel(fdTarget)) {
+    send441UserNotInChannel(client, targetNick, channel);
+    return;
+  }
+  broadcastInChannel(*targetClient, channel, "KICK", reason);
```

Comment:
I think the KICK message has wrong format.

Server sends
`:yusengok KICK #chan2 :There is no reason :)`
Hexchat prints
` yusengok has kicked :There from #chan2 (is no reason :))`

Right format is
`:<Client's nick who sends kick command> KICK #channelName :<Kicked client's nick> <Reason>`

Update code like this:
```
std::string message = targetNick + " " + reason + "\r\n"; 
broadcastInChannel(client, channel, "KICK", message)
```

The server sends
`:lala KICK #42 :yusengok C'est la vie !`
Hexchat prints
`lala has kicked :yusengok from #42 (C'est la vie !)`